src/org/sosy_lab/java_smt/solvers/z3/Z3IntegerFormulaManager.java

```diff
@@ -31,9 +31,58 @@ protected Long makeNumberImpl(double pNumber) {
     return makeNumberImpl((long) pNumber);
   }
 
+  /**
+   * Creates an integer formula from a BigDecimal value. For Z3, we need to handle this specially to
+   * avoid segfaults when dealing with decimal values that have fractional parts.
+   *
+   * <p>The issue is that the default implementation tries to represent decimal values as division
+   * operations, which can cause segfaults in Z3 when used with integer sorts.
+   *
+   * <p>This method safely converts BigDecimal values by:
+   * <ol>
+   *   <li>Using exact conversion for integers (no fractional part)</li>
+   *   <li>Using euclidean division for values with fractional parts to maintain consistency
+   *       with other solvers in JavaSMT</li>
+   * </ol>
+   *
+   * <p>This prevents the segfault described in issue #457.
+   */
   @Override
   protected Long makeNumberImpl(BigDecimal pNumber) {
-    return decimalAsInteger(pNumber);
+    if (pNumber.scale() <= 0) {
+      try {
+        return makeNumberImpl(pNumber.toBigIntegerExact());
+      } catch (ArithmeticException e) {
+        // This shouldn't happen since we checked scale <= 0
+        throw new AssertionError("Unexpected error converting BigDecimal", e);
+      }
+    } else {
+      // For values with fractional parts, use euclidean division to maintain
+      // consistency with the behavior of other solvers in JavaSMT
+      BigInteger unscaledValue = pNumber.unscaledValue();
+      BigInteger scale = BigInteger.TEN.pow(pNumber.scale());
+      return makeNumberImpl(euclideanDivision(unscaledValue, scale));
+    }
```

Comment:
EuclideanDivision looks like a complex way of computing the following, which might be more readable:
```
roundingmode = pNumber.compareTo(BigDecimal.ZERO) < 0 ? BigDecimal.ROUND_CEILING : BigDecimal.ROUND_FLOOR;
num = pNumber.setScale(0, roundingMode).toBigInteger();
return makeNumberImpl(num);
```

Reply:
Simplified in e5d1c16fe0d48965924014b7d5125f6c46d3a627